SNLS_NLSPowell.h

```diff
@@ -0,0 +1,729 @@
+                        diag[i] = work_arr2[i];
+                    } else {
+                        diag[i] = 1.0;
+                    }
+                }
+            }
+
+            m_xnorm = snls::linalg::norm<m_nDim>(m_x);
+            m_delta = m_factor * m_xnorm;
+            if (m_delta == 0.0) {
+                m_delta = m_factor; 
+            }
+        }
+
+        if (scaleDiag) {
+            for (int i = 0; i < m_nDim; i++) {
+                diag[i] = (diag[i] > work_arr2[i]) ? diag[i] : work_arr2[i];
+            }
+        }
+
+        // Jacobian is our R matrix and Q
+        snls::houseHolderQR<m_nDim, m_nDim>(Jacobian, Q, work_arr1, work_arr2, work_arr3);
+
+        snls::linalg::matTVecMult<m_nDim, m_nDim>(Q, residual, qtf);
+
+        // std::cout << "R" << std::endl;
+        // snls::linalg::printMat<m_nDim>(Jacobian);
+        // std::cout << "Q" << std::endl;
+        // snls::linalg::printMat<m_nDim>(Q);
+        // std::cout << "qtf" << std::endl;
+        // snls::linalg::printVec<m_nDim>(qtf);
+        // std::cout << "diag" << std::endl;
+        // snls::linalg::printVec<m_nDim>(diag);
+
+        for (int i = 0; i < m_nDim; i++) {
+            work_arr1[i] = m_x[i];
+        }
+
+        for (;;) {
+            // Determine the direction p
+            this->dogleg(work_arr1, Jacobian, diag, qtf, m_delta);
+            // Store the direction of p and x + p
+            double pnorm = 0.0;
+            for (int i = 0; i < m_nDim; i++) {
+                work_arr1[i] *= -1.0; 
+                // If our solution isn't satisfactory later on we'll
+                // reject the solution and calculate the new value of things
+                m_x[i] += work_arr1[i];
+                pnorm += (diag[i] * work_arr1[i]) * (diag[i] * work_arr1[i]);
+            }
+            pnorm = sqrt(pnorm);
+
+            // std::cout << "work_arr1" << std::endl;
+            // snls::linalg::printVec<m_nDim>(work_arr1);
+            // std::cout << "m_x" << std::endl;
+            // snls::linalg::printVec<m_nDim>(m_x);
+
+
+            // On the first iteration adjust the initial step bound
+            if (m_nIters == 1) {
+                m_delta = fmin(m_delta, pnorm);
+            }
+
+            // Evaluate the function at x + p and calculate its norm
+            // Previously iterations kept x and residual seperated by
+            // making use of working arrays...
+            // This is a waste of memory as we can just use residual and x arrays.
+            // Also, we currently don't check on the success of this...
+            // and we should at this back in at some point
+            // However in the mean time, this is to quite the compiler
+            // bool rjSuccess = 
+            this->computeRJ(residual, nullptr);
+            // Need to think about what to do if this fails as things
+            // are handled differently from the TrDL solver.
+            // if ( !(rjSuccess) ) {
```

Comment:
Repeated typos in comments, and the need to do _something_ if computeRJ comes back with an error. My guess is that you just go to SNLS_FAIL if computeRJ has an error, at least for now. You can code a cutback in the step size or the like later.

Reply:
Yeah I wasn't too sure how best to handle the cut back portion. I can just add a SNLS_FAIL in the mean time until we figure that issue out.

Reply:
This has been fixed with latest revisions of things.